File: 03-Stacks/20250728-intermediate/go/solution_test.go
package main

import (
	"testing"
)

func TestBasicMinStackOperations(t *testing.T) {
	minStack := NewMinStack()
	
	// Push operations
	minStack.Push(3)
	minStack.Push(1)  
	minStack.Push(4)
	
	// Test GetMin
	min, err := minStack.GetMin()
	if err != nil {
		t.Errorf("GetMin() failed: %v", err)
	}
	if min != 1 {
		t.Errorf("GetMin() = %d, want 1", min)
	}
	
	// Test Peek
	top, err := minStack.Peek()
	if err != nil {
		t.Errorf("Peek() failed: %v", err)
	}
	if top != 4 {
		t.Errorf("Peek() = %d, want 4", top)
	}
	
	// Test Size
	if size := minStack.Size(); size != 3 {
		t.Errorf("Size() = %d, want 3", size)
	}
}

func TestPopAndMinUpdate(t *testing.T) {
	minStack := NewMinStack()
	
	minStack.Push(5)
	minStack.Push(2)
	minStack.Push(7)
	
	// Pop operation
	popped, err := minStack.Pop()
	if err != nil {
		t.Errorf("Pop() failed: %v", err)
	}
	if popped != 7 {
		t.Errorf("Pop() = %d, want 7", popped)
	}
	
	// Check min after pop
	min, err := minStack.GetMin()
	if err != nil {
		t.Errorf("GetMin() failed: %v", err)
	}
	if min != 2 {
		t.Errorf("GetMin() after pop = %d, want 2", min)
	}
}

func TestMultipleSameMinimumValues(t *testing.T) {
	minStack := NewMinStack()
	
	minStack.Push(2)
	minStack.Push(3)
	minStack.Push(2)
	minStack.Push(1)
	minStack.Push(2)
	
	// Pop the top 2
	popped, err := minStack.Pop()
	if err != nil {
		t.Errorf("Pop() failed: %v", err)
	}
	if popped != 2 {
		t.Errorf("Pop() = %d, want 2", popped)
	}
	
	// Min should still be 1
	min, err := minStack.GetMin()
	if err != nil {
		t.Errorf("GetMin() failed: %v", err)
	}
	if min != 1 {
		t.Errorf("GetMin() after pop = %d, want 1", min)
	}
}

func TestBasicUndoPushOperation(t *testing.T) {
	minStack := NewMinStack()
	
	minStack.Push(10)
	minStack.Push(5)
	
	// Undo the last push
	success := minStack.Undo()
	if !success {
		t.Error("Undo() should return true")
	}
	
	// Check result
	top, err := minStack.Peek()
	if err != nil {
		t.Errorf("Peek() failed: %v", err)
	}
	if top != 10 {
		t.Errorf("Peek() after undo = %d, want 10", top)
	}
	
	if size := minStack.Size(); size != 1 {
		t.Errorf("Size() after undo = %d, want 1", size)
	}
}

func TestBasicUndoPopOperation(t *testing.T) {
	minStack := NewMinStack()
	
	minStack.Push(1)
	minStack.Push(2)
	minStack.Pop()
	
	// Undo the pop
	success := minStack.Undo()
	if !success {
		t.Error("Undo() should return true")
	}
	
	// Check result
	top, err := minStack.Peek()
	if err != nil {
		t.Errorf("Peek() failed: %v", err)
	}
	if top != 2 {
		t.Errorf("Peek() after undo = %d, want 2", top)
	}
	
	if size := minStack.Size(); size != 2 {
		t.Errorf("Size() after undo = %d, want 2", size)
	}
}

func TestMultipleUndoOperations(t *testing.T) {
	minStack := NewMinStack()
	
	minStack.Push(1)
	minStack.Push(2)
	minStack.Push(3)
	
	// Undo twice
	minStack.Undo()
	minStack.Undo()
	
	// Check result
	top, err := minStack.Peek()
	if err != nil {
		t.Errorf("Peek() failed: %v", err)
	}
	if top != 1 {
		t.Errorf("Peek() after undos = %d, want 1", top)
	}
	
	min, err := minStack.GetMin()
	if err != nil {
		t.Errorf("GetMin() failed: %v", err)
	}
	if min != 1 {
		t.Errorf("GetMin() after undos = %d, want 1", min)
	}
}

func TestUndoOnEmptyHistory(t *testing.T) {
	minStack := NewMinStack()
	
	// Try to undo with no history
	success := minStack.Undo()
	if success {
		t.Error("Undo() should return false when no history")
	}
	
	// Stack should remain empty
	if !minStack.IsEmpty() {
		t.Error("Stack should remain empty after failed undo")
	}
}

func TestHistoryTracking(t *testing.T) {
	minStack := NewMinStack()
	
	minStack.Push(5)
	minStack.Push(3)
	minStack.Pop()
	minStack.Push(7)
	
	history := minStack.GetHistory()
	expected := []string{"Push(5)", "Push(3)", "Pop()", "Push(7)"}
	
	if len(history) != len(expected) {
		t.Errorf("History length = %d, want %d", len(history), len(expected))
	}
	
	for i, op := range expected {
		if i >= len(history) || history[i] != op {
			t.Errorf("History[%d] = %s, want %s", i, history[i], op)
		}
	}
}

func TestClearHistoryFunctionality(t *testing.T) {
	minStack := NewMinStack()
	
	minStack.Push(1)
	minStack.Push(2)
	minStack.Pop()
	
	// Clear history
	minStack.ClearHistory()
	
	// History should be empty
	history := minStack.GetHistory()
	if len(history) != 0 {
		t.Errorf("History length after clear = %d, want 0", len(history))
	}
	
	// But stack content should be unchanged
	if size := minStack.Size(); size != 1 {
		t.Errorf("Size() after clear history = %d, want 1", size)
	}
	
	top, err := minStack.Peek()
	if err != nil {
		t.Errorf("Peek() failed: %v", err)
	}
	if top != 1 {
		t.Errorf("Peek() after clear history = %d, want 1", top)
	}
}

func TestComplexMixedOperationsWithUndo(t *testing.T) {
	minStack := NewMinStack()
	
	minStack.Push(4)
	minStack.Push(1)
	minStack.Push(6)
	
	// Pop
	minStack.Pop()
	// Stack: [4, 1], min: 1
	min1, _ := minStack.GetMin()
	if min1 != 1 {
		t.Errorf("GetMin() after pop = %d, want 1", min1)
	}
	
	// Undo the pop
	minStack.Undo()
	// Stack: [4, 1, 6], min: 1
	min2, _ := minStack.GetMin()
	if min2 != 1 {
		t.Errorf("GetMin() after undo = %d, want 1", min2)
	}
	
	// Pop again
	minStack.Pop()
	// Stack: [4, 1], min: 1
	min3, _ := minStack.GetMin()
	if min3 != 1 {
		t.Errorf("GetMin() after second pop = %d, want 1", min3)
	}
	
	// Undo the pop again
	minStack.Undo()
	// Stack: [4, 1, 6], min: 1
	min4, _ := minStack.GetMin()
	if min4 != 1 {
		t.Errorf("GetMin() after second undo = %d, want 1", min4)
	}
}

File: 03-Stacks/20250728-intermediate/ts/solution.test.ts
import { MinStack } from './solution';

describe('MinStack', () => {
  test('Test Case 1: Basic MinStack Operations', () => {
    const minStack = new MinStack();
    
    // Push operations
    minStack.push(3);
    minStack.push(1);
    minStack.push(4);
    
    // Test GetMin
    expect(minStack.getMin()).toBe(1);
    
    // Test Peek
    expect(minStack.peek()).toBe(4);
    
    // Test Size
    expect(minStack.size()).toBe(3);
  });

  test('Test Case 2: Pop and Min Update', () => {
    const minStack = new MinStack();
    
    minStack.push(5);
    minStack.push(2);
    minStack.push(7);
    
    // Pop operation
    expect(minStack.pop()).toBe(7);
    
    // Check min after pop
    expect(minStack.getMin()).toBe(2);
  });

  test('Test Case 3: Multiple Same Minimum Values', () => {
    const minStack = new MinStack();
    
    minStack.push(2);
    minStack.push(3);
    minStack.push(2);
    minStack.push(1);
    minStack.push(2);
    
    // Pop the top 2
    expect(minStack.pop()).toBe(2);
    
    // Min should still be 1
    expect(minStack.getMin()).toBe(1);
  });

  test('Test Case 4: Basic Undo Push Operation', () => {
    const minStack = new MinStack();
    
    minStack.push(10);
    minStack.push(5);
    
    // Undo the last push
    expect(minStack.undo()).toBe(true);
    
    // Check result
    expect(minStack.peek()).toBe(10);
    expect(minStack.size()).toBe(1);
  });

  test('Test Case 5: Basic Undo Pop Operation', () => {
    const minStack = new MinStack();
    
    minStack.push(1);
    minStack.push(2);
    minStack.pop();
    
    // Undo the pop
    expect(minStack.undo()).toBe(true);
    
    // Check result
    expect(minStack.peek()).toBe(2);
    expect(minStack.size()).toBe(2);
  });

  test('Test Case 6: Multiple Undo Operations', () => {
    const minStack = new MinStack();
    
    minStack.push(1);
    minStack.push(2);
    minStack.push(3);
    
    // Undo twice
    minStack.undo();
    minStack.undo();
    
    // Check result
    expect(minStack.peek()).toBe(1);
    expect(minStack.getMin()).toBe(1);
  });

  test('Test Case 7: Undo on Empty History', () => {
    const minStack = new MinStack();
    
    // Try to undo with no history
    expect(minStack.undo()).toBe(false);
    
    // Stack should remain empty
    expect(minStack.isEmpty()).toBe(true);
  });

  test('Test Case 8: History Tracking', () => {
    const minStack = new MinStack();
    
    minStack.push(5);
    minStack.push(3);
    minStack.pop();
    minStack.push(7);
    
    const history = minStack.getHistory();
    const expected = ['Push(5)', 'Push(3)', 'Pop()', 'Push(7)'];
    
    expect(history).toEqual(expected);
  });

  test('Test Case 9: Clear History Functionality', () => {
    const minStack = new MinStack();
    
    minStack.push(1);
    minStack.push(2);
    minStack.pop();
    
    // Clear history
    minStack.clearHistory();
    
    // History should be empty
    expect(minStack.getHistory()).toEqual([]);
    
    // But stack content should be unchanged
    expect(minStack.size()).toBe(1);
    expect(minStack.peek()).toBe(1);
  });

  test('Test Case 10: Complex Mixed Operations with Undo', () => {
    const minStack = new MinStack();
    
    minStack.push(4);
    minStack.push(1);
    minStack.push(6);
    
    // Pop
    minStack.pop();
    // Stack: [4, 1], min: 1
    expect(minStack.getMin()).toBe(1);
    
    // Undo the pop
    minStack.undo();
    // Stack: [4, 1, 6], min: 1
    expect(minStack.getMin()).toBe(1);
    
    // Pop again
    minStack.pop();
    // Stack: [4, 1], min: 1
    expect(minStack.getMin()).toBe(1);
    
    // Undo the pop again
    minStack.undo();
    // Stack: [4, 1, 6], min: 1
    expect(minStack.getMin()).toBe(1);
  });

  test('Error Cases: Empty Stack Operations', () => {
    const minStack = new MinStack();
    
    // Pop from empty stack
    expect(() => minStack.pop()).toThrow('Stack is empty');
    
    // Peek at empty stack
    expect(() => minStack.peek()).toThrow('Stack is empty');
    
    // GetMin on empty stack
    expect(() => minStack.getMin()).toThrow('Stack is empty');
  });
});